### PR TITLE
Feature/implement comment thread subscriber

### LINKED ISSUE #5195: Stop sending emails to everyone in the organization for comments
When a new comment is added, everyone in the organization gets an email. We need to change this behavior as per the following:

- When a new comment is added without tagging anyone, only existing users in the thread who have commented in the thread will get an email. So if user1, user2 and user3 have put a comment in a thread and there is a new one from user4, then only 1, 2 and 3 will get a notification through email
- When a new comment is added by tagging one or more users, then the tagged users will receive notifications mentioning that they are tagged. Also the other users in the thread will get a notification about the message - just like the previous case
- When an user leaves or is removed from an organization, they'll no more receive any notifications.

Here is a sample user story:

1. U1 created a new thread T1 and adds a new comment C1
No notifications to anyone
2. U2 added C2 in T1
U1 is notified
3. U3 added C3 in T1
U1 and U2 are notified
4. U3 adds C4 in T1 mentioning U4
U1 and U2 are notified
U4 is notified
5. U2 adds C5 mentioning no one
U1, U3 are notified. (the small change here the user U4  should be notified with all comments once after he had been Tagged in T1)

### PR DESCRIPTION
**Pull request by nayan-rafiq**
## Description
Currently, if there is a new comment - an email and a notification is sent to everyone in the organization. For an organization with many users and applications - this will be very annoying. Instead we should send email to interested people only.

This PR will send notification and email to the users who have participated in the thread only or tagged by others in this thread. In future, we'll be able to add feature to unsubscribe from the thread notifications also.

Fixes #5195 

## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- Unit tests
- Manual test on developer machine

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java [hunk 1]
```diff
@@ -31,6 +31,35 @@ public static boolean isUserMentioned(Comment comment, String userEmail) {
         return false;
     }
 
+    /**
+     * Returns the list of usernames who should subscribe to a thread
+     * It'll include the author username. It'll also include anyone who is mentioned in this comment by the author.
+     * For example, if this comment is from user1 and user2 and user3 are mentioned in this comment,
```
Comment (abhvsn) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java:37]: Can we modify comment from _For example, if this comment is from user1 and user2 and user3 are mentioned in this comment_ to __For example if this comment is from user1 where user2 and user3 are mentioned in this comment__

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java [hunk 2]
```diff
@@ -31,6 +31,35 @@ public static boolean isUserMentioned(Comment comment, String userEmail) {
...
+    public static List<String> getSubscriberUsernames(Comment comment) {
+        List<String> usernameList = new ArrayList<>();
+        // add the author itself
+        usernameList.add(comment.getAuthorUsername());
+
+        if(comment.getBody() != null && comment.getBody().getEntityMap() != null) {
+            for(String key : comment.getBody().getEntityMap().keySet()) {
+                Comment.Entity commentEntity = comment.getBody().getEntityMap().get(key);
+                if(commentEntity != null && commentEntity.getType() != null
+                        && commentEntity.getType().equals("mention")) {
```
Comment (abhvsn) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java:51]: Instead of "mention" can we use Mention.class.toString()?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java [hunk 3]
```diff
@@ -31,6 +31,35 @@ public static boolean isUserMentioned(Comment comment, String userEmail) {
...
+                    // this comment has a mention, check the provided user is mentioned or not
+                    if(commentEntity.getData() != null) {
```
Comment (abhvsn) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java:53]: We have already checked the null condition in the parent if statement, is it still required?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java (hunk 2 above)
Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java:51]: If we do that, we probably need `.equalsIgnoreCase` instead of `.equals` 😄

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java [hunk 4]
```diff
@@ -31,6 +31,35 @@ public static boolean isUserMentioned(Comment comment, String userEmail) {
...
+    /**
+     * Returns the list of usernames who should subscribe to a thread
+     * It'll include the author username. It'll also include anyone who is mentioned in this comment by the author.
+     * For example, if this comment is from user1 and user2 and user3 are mentioned in this comment,
+     * it'll return a list of user1, user2, user3
+     * @param comment The comment object
+     * @return list of usernames. Size of the list will be at least 1
+     */
+    public static List<String> getSubscriberUsernames(Comment comment) {
+        List<String> usernameList = new ArrayList<>();
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java:43]: Since it looks like we are treating the collections of subscribers as a _set_, should we be consistent and treat it as a `Set` everywhere? Unless there's a specific need to use `List` here?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/CommentServiceImpl.java [hunk 5]
```diff
@@ -235,20 +248,9 @@ public CommentServiceImpl(
                 .flatMap(tuple -> {
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/CommentServiceImpl.java:248]: Since we don't have any reactive code in this inner function anymore, we can just use `.map` instead, and change the `Mono.just(commentThread)` to just `commentThread`.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/EmailEventHandler.java [hunk 6]
```diff
@@ -67,8 +76,8 @@
     @EventListener
     public void handle(CommentAddedEvent event) {
         this.sendEmailForComment(
-                event.getAuthorUserName(), event.getApplication(), event.getOrganization(),
-                event.getComment(), event.getOriginHeader()
+                event.getAuthorUserName(), event.getOrganization(),
+                event.getComment(), event.getOriginHeader(), event.getSubscribers()
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/EmailEventHandler.java:80]: A small request, since I'm seeing this pattern in a lot of places, if we are splitting a list of arguments over multiple lines, can we change it to having one argument per line? So, either we have all arguments on the same line, or one argument per line. Otherwise it's really hard to follow when reading/reviewing code because then I'm essentially counting arguments per line. Oh the first line has 2 and the second line has 3... etc. This applies to both calling methods and defining methods. Just a small thing that can make reading/reviewing much simpler. Thanks!

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/EmailEventHandler.java [hunk 7]
```diff
@@ -41,18 +40,28 @@
     private final OrganizationRepository organizationRepository;
     private final ApplicationRepository applicationRepository;
 
-    public Mono<Boolean> publish(String authorUserName, String applicationId, Comment comment, String originHeader) {
+    public Mono<Boolean> publish(String authorUserName, String applicationId, Comment comment, String originHeader,
+                                 Set<String> subscribers) {
+        if(subscribers == null || subscribers.size() == 0) {  // no subscriber found, return without doing anything
```
Comment (abhvsn) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/EmailEventHandler.java:45]: Instead of checking size() can we use isEmpty() here?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/EmailEventHandler.java [hunk 8]
```diff
@@ -41,18 +40,28 @@
...
-                        new CommentAddedEvent(authorUserName, organization, application, originHeader, comment)
+                        new CommentAddedEvent(authorUserName, organization, application, originHeader, comment, subscribers)
                 );
                 return Mono.just(organization);
             });
         }).thenReturn(Boolean.TRUE);
     }
 
     public Mono<Boolean> publish(String authorUserName, String applicationId, CommentThread thread, String originHeader) {
+        if(thread.getSubscribers() == null || thread.getSubscribers().size() == 0) {
```
Comment (abhvsn) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/EmailEventHandler.java:60]: Same as above

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/EmailEventHandler.java (hunk 7 above)
Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/EmailEventHandler.java:45]: Even better, there's `CollectionUtils.isEmpty()` which will do the `null` check also for us, so I tend to recommend that more often. :)

**Comment by sharat87** [comment]
@nayan-rafiq This looks good dude. I'm honestly surprised that adding subscribers is actually this small a change. Nicely done.

Tangential, I noticed that we save the application name as part of comment thread objects, right? So, I'm guessing we need to update this when an application gets renamed no? Not high priority, but it's something we need to take up or it'll be raised by a confused user someday 😉

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java (hunk 1 above)
Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java:37]: Good point. I'll update accordingly.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java (hunk 2 above)
Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java:51]: @abhvsn `type` is set from front-end. Changing it to class name may break the code if we change the class name.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java (hunk 3 above)
Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java:53]: Yes, because an entity has data as an inner class. As it's sent from front end, it may produce NPE if entity has no data. That's why I added this check just to be safe.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java (hunk 4 above)
Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CommentUtils.java:43]: Good point, changing it to `Set` here also.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/EmailEventHandler.java (hunk 6 above)
Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/EmailEventHandler.java:80]: Agreed. Will change it accordingly.

**Comment by nayan-rafiq** [comment]
@sharat87 @abhvsn Thanks a lot for your valuable comments. I've updated the PR as per our discussions. Would you please have a look again?

**Comment by sharat87** [comment]
Looks ready! Please do a squash merge.

**Review by sharat87 — APPROVED** [review]
